src/test/java/com/puppycrawl/tools/checkstyle/PropertyCacheFileTest.java

```diff
@@ -284,4 +295,133 @@ public void testChangeInConfig() throws Exception {
+        }
+    }
+
+    /**
+     * This test invokes {@code putExternalResources} twice to invalidate cache.
+     * And asserts that two different exceptions produces different content,
+     * but two exceptions with same message produces one shared content.
+     *
+     * @param rawMessages exception messages separated by ';'
+     */
+    @ParameterizedTest
+    @ValueSource(strings = {"Same;Same", "First;Second"})
+    public void testPutNonExistentExternalResource(String rawMessages) throws Exception {
+        final File cacheFile = File.createTempFile("junit", null, temporaryFolder);
+        final String[] messages = rawMessages.split(";");
+        // We mock getUriByFilename method of CommonUtil to guarantee that it will
+        // throw CheckstyleException with the specific content.
+        try (MockedStatic<CommonUtil> commonUtil = mockStatic(CommonUtil.class)) {
+            final int numberOfRuns = messages.length;
+            final String[] configHashes = new String[numberOfRuns];
+            final String[] externalResourceHashes = new String[numberOfRuns];
+            for (int i = 0; i < numberOfRuns; i++) {
+                commonUtil.when(() -> CommonUtil.getUriByFilename(any(String.class)))
+                        .thenThrow(new CheckstyleException(messages[i]));
+                final Configuration config = new DefaultConfiguration("myConfig");
+                final PropertyCacheFile cache = new PropertyCacheFile(config, cacheFile.getPath());
+                cache.load();
+
+                configHashes[i] = cache.get(PropertyCacheFile.CONFIG_HASH_KEY);
+                assertWithMessage("Config hash key should not be null")
+                        .that(configHashes[i])
+                        .isNotNull();
+
+                final Set<String> nonExistentExternalResources = new HashSet<>();
+                final String externalResourceFileName = "non_existent_file.xml";
+                nonExistentExternalResources.add(externalResourceFileName);
+                cache.putExternalResources(nonExistentExternalResources);
+
+                externalResourceHashes[i] = cache.get(PropertyCacheFile.EXTERNAL_RESOURCE_KEY_PREFIX
+                        + externalResourceFileName);
+                assertWithMessage("External resource hashes should not be null")
+                        .that(externalResourceHashes[i])
+                        .isNotNull();
+
+                cache.persist();
+
+                final Properties cacheDetails = new Properties();
+                try (BufferedReader reader = Files.newBufferedReader(cacheFile.toPath())) {
+                    cacheDetails.load(reader);
+                }
+
+                final int expectedNumberOfObjectsInCacheFile = 2;
+                assertWithMessage("Unexpected number of objects in cache")
+                        .that(cacheDetails)
+                        .hasSize(expectedNumberOfObjectsInCacheFile);
```

Comment:
better to inline expectedNumberOfObjectsInCacheFile to assert.

Reply:
done